File: frontend/src/types/business.ts
export type Platform = 'WEIBO' | 'XIAOHONGSHU' | 'WECHAT_OFFICIAL' | 'TOUTIAO' | 'X'

export interface Article {
  id: number
  title: string
  sourceText: string
  summary?: string
  topic?: string
  targetAudience?: string
  tone?: string
  keywords: string[]
  status: string
  variantCount: number
  coverThumbnailUrl?: string
  createdAt: string
  updatedAt: string
  variants?: Variant[]
}

export interface ArticleEngagementMetric {
  scheduleId: number
  platform: Platform
  metricDate: string
  impressions: number
  likes: number
  comments: number
  collects: number
  shares: number
  followers: number
  engagementTotal: number
  engagementRate: number
  dataSource: string
}

export interface ArticleEngagementTotals {
  impressions: number
  likes: number
  comments: number
  collects: number
  shares: number
  engagementTotal: number
  engagementRate: number
}

export interface ArticleEngagementSummary {
  articleId: number
  hasData: boolean
  platforms?: ArticleEngagementMetric[]
  totals: ArticleEngagementTotals | null
  dataSources: string[]
  simulated: boolean
}

export interface Variant {
  id: number
  articleId: number
  platform: Platform
  versionNo: number
  title: string
  contentText: string
  contentHtml?: string
  formatProfileJson?: WechatFormatProfile
  hashtagsJson: string[]
  wordCount: number
  modelName: string
  promptVersion: string
  generationDurationMs: number
  tokenUsage: number
  promptTokens: number
  completionTokens: number
  estimatedCost: number
  qualityScore: number
  manualEditRatio: number
  reviewStatus: string
  reviewDetailJson?: Record<string, unknown>
  createdAt: string
}

export interface WechatFormatProfile {
  theme: 'clean' | 'brand' | 'editorial'
  accent_color: string
  font_size: number
  line_height: number
  paragraph_spacing: number
  first_line_indent: boolean
  link_footnotes: boolean
}

export interface WechatThemeProfile {
  key: WechatFormatProfile['theme']
  name: string
  description: string
  accent_color: string
  heading_style: string
  quote_background: string
}

export type GenerationPlatformStatus = 'PENDING' | 'RUNNING' | 'RETRYING' | 'SUCCESS' | 'FAILED'

export interface DeepCreationCandidate {
  title: string
  content: string
  hashtags: string[]
  warnings?: string[]
  summary?: string
  cover_text?: string
  cover_prompt?: string
  author?: string
}

export interface GenerationPlatformProgress {
  status: GenerationPlatformStatus
  progress: number
  stage?: string
  message?: string
  attempt: number
  maxAttempts?: number
  variantId?: number
  error?: string
  durationMs: number
  tokenUsage: number
  updatedAt?: string
  brief?: Record<string, unknown>
  strategy?: {
    angle: string
    hook: string
    reader_value: string
    structure: string[]
    cta: string
  }
  review?: Record<string, unknown>
  candidateTitles?: string[]
  candidates?: DeepCreationCandidate[]
  selectedCandidate?: number
  userSelectedCandidate?: number
  userSelectedVariantId?: number
}

export interface GenerationTask {
  id: string
  articleId: number
  status: 'PENDING' | 'RUNNING' | 'SUCCESS' | 'PARTIAL_SUCCESS' | 'FAILED'
  progress: number
  platformsJson: Platform[]
  platformStatusJson: Partial<Record<Platform, GenerationPlatformProgress>>
  resultVariantIdsJson: number[]
  variants: Variant[]
  provider: string
  modelName: string
  promptVersion: string
  tokenUsage: number
  durationMs: number
  errorMessage?: string
  createdAt?: string
  updatedAt?: string
  optionsJson?: Record<string, unknown>
}

export interface TrendItem {
  id: string
  source: 'BAIDU' | 'HACKER_NEWS'
  sourceName: string
  rank: number
  title: string
  summary: string
  url: string
  imageUrl?: string
  heat?: number
  heatLabel: string
  fetchedAt: string
  tags: string[]
}

export interface TrendAngle {
  title: string
  audience: string
  hook: string
  outline: string[]
  creative_goal: string
}

export interface TrendAnalysis {
  relevance_reason: string
  recommended_angle_index: number
  angles: TrendAngle[]
  risk_notes: string[]
  verification_questions: string[]
  provider: string
  modelName: string
  tokenUsage: number
}

export interface MediaAsset {
  id: number
  articleId?: number
  articleTitle?: string
  variantId?: number
  imageUrl: string
  thumbnailUrl: string
  altText?: string
  usageType: 'COVER' | 'BODY'
  source: string
  selected: boolean
  title?: string
  collection?: string
  tags: string[]
  favorite: boolean
  licenseType?: string
  licenseNote?: string
  photographerName?: string
  photographerUrl?: string
  createdAt: string
}

export interface ResearchItem {
  id: number
  title: string
  summary?: string
  url?: string
  imageUrl?: string
  source: string
  sourceName?: string
  sourceId?: string
  topicCluster?: string
  tags: string[]
  notes?: string
  status: 'INBOX' | 'RESEARCHING' | 'READY' | 'USED'
  archived: boolean
  createdAt: string
  updatedAt: string
}

export interface ScheduleBacklogItem {
  articleId: number
  variantId: number
  articleTitle: string
  variantTitle: string
  platform: Platform
  reviewStatus: string
  coverReady: boolean
  accountReady: boolean
  ready: boolean
  blockers: string[]
  updatedAt: string
}

export interface OperationStep {
  key: string
  name: string
  status: string
  stage?: string
  message?: string
  progress: number
  durationMs: number
  tokenUsage?: number
  error?: string
  attempt?: number
  createdAt?: string
}

export interface OperationRun {
  id: string
  sourceId: string | number
  type: 'GENERATION' | 'PUBLISH'
  title: string
  subtitle: string
  status: string
  progress: number
  provider?: string
  modelName?: string
  tokenUsage?: number
  durationMs: number
  errorMessage?: string
  createdAt: string
  updatedAt: string
  scheduledAt?: string
  retryCount?: number
  maxRetryCount?: number
  steps: OperationStep[]
}

export type TimeSource = 'RECOMMENDED' | 'ALTERNATIVE' | 'CUSTOM'

export interface Schedule {
  id: number
  articleId: number
  variantId: number
  platform: Platform
  articleTitle: string
  variantTitle: string
  scheduledAt: string
  accountId: number
  accountName: string
  publishMode: PublishMode
  status: string
  retryCount: number
  actualPublishAt?: string
  publishedUrl?: string
  externalId?: string
  resultMode?: string
  publishPackageJson?: PublishPackage
  errorMessage?: string
  logs?: Array<Record<string, unknown>>
  recommendationId?: number
  recommendedAt?: string
  timeSource: TimeSource
  contentType?: string
  contentTypeName?: string
  timeDeviationMinutes?: number | null
  usedRecommendedTime?: boolean
  recommendationSnapshotJson?: Record<string, unknown>
}

export interface RecommendationReason {
  type: string
  description: string
  contribution: number
}

export interface RecommendationAlternative {
  recommendedAt: string
  score: number
  confidence: string
  sampleCount: number
  reason: string
}

export interface ScheduleConflict {
  scheduleId: number
  articleTitle?: string
  scheduledAt: string
  minutes: number
  sameAccount: boolean
  level: 'CONFLICT' | 'DENSITY'
  message: string
}

export type TimeWindow = '30D' | '90D' | 'ALL' | 'CUSTOM'

export interface TimeWindowInfo {
  window: TimeWindow
  label: string
  startDate: string | null
  endDate: string | null
  sufficient: boolean
  message: string
  sampleCount?: number
  accountSampleCount?: number
  baselineSampleCount?: number
}

export interface ActivityCurvePoint {
  hour: number
  time: string
  platformPrior: number
  accountHistory: number | null
  publicBaseline: number | null
  sampleCount: number
}

export interface PublishTimeRecommendation {
  id: number
  articleId: number
  variantId?: number
  platform: Platform
  accountId?: number
  recommendedAt: string
  score: number
  confidence: 'HIGH' | 'MEDIUM' | 'LOW'
  algorithmVersion: string
  contentType?: string
  contentTypeName?: string
  contentTypeProvider?: string
  topic?: string
  audience?: string
  sampleCount: number
  accountSampleCount: number
  baselineSampleCount: number
  reasons: RecommendationReason[]
  alternatives: RecommendationAlternative[]
  warnings: string[]
  conflicts: ScheduleConflict[]
  weights: { baseline: number; history: number; content: number; timezone: number }
  dataSource: {
    baseline: string
    accountHistory: string
    priorRuleCount: number
    publicSampleCount: number
    accountSampleCount: number
    sourceTypes: Array<{ sourceType: string; label: string; count: number }>
  }
  dataSufficiency: {
    level: string
    accountSamples: number
    slotSamples: number
    sufficient: boolean
    message: string
  }
  narrative?: string
  narrativeProvider?: string
  window?: TimeWindowInfo
  curve: ActivityCurvePoint[]
}

export interface ActivityBucket {
  sampleCount: number
  score: number
  engagementRate: number
  avgViews: number
  avgLikes: number
  avgComments: number
  avgShares: number
  avgFavorites: number
}

export interface ExperimentSampleRow {
  id: number
  experimentId: number
  scheduleId?: number
  groupType: 'CONTROL' | 'TREATMENT'
  sampleLabel: string
  metricValueJson: Record<string, unknown>
  assignmentSource?: 'AUTO' | 'MANUAL'
  assignmentReason?: string | null
  createdAt: string
}

export interface ActivityAnalysis {
  filters: Record<string, unknown>
  windowInfo?: TimeWindowInfo
  sampleCount: number
  accountSampleCount: number
  baselineSampleCount: number
  globalMeanScore: number
  weekday: Array<ActivityBucket & { dayOfWeek: number; name: string }>
  hourly: Array<ActivityBucket & { hour: number; time: string }>
  heatmap: Array<{
    dayOfWeek: number
    name: string
    hour: number
    sampleCount: number
    score: number
  }>
  contentTypes: Array<
    ActivityBucket & {
      contentType: string
      contentTypeName: string
      bestSlots: Array<{
        dayOfWeek: number
        dayName: string
        hour: number
        time: string
        sampleCount: number
        score: number
      }>
    }
  >
  topSlots: Array<{
    dayOfWeek: number
    dayName: string
    hour: number
    time: string
    sampleCount: number
    score: number
    rawScore: number
  }>
  completeness: {
    score: number
    fields: Array<{
      field: string
      label: string
      filled: number
      total: number
      percent: number
    }>
    dateRange: { start: string | null; end: string | null }
  }
  sources: Array<{ sourceType: string; label: string; count: number }>
  scoreFormula: string
  notice: string
}

export interface HistoryImportPreview {
  filename: string
  headers: string[]
  mapping: Record<string, string>
  fields: Array<{ field: string; label: string; required: boolean; type: string }>
  totalRows: number
  validRows: number
  errorRows: number
  duplicateInFile: number
  duplicateInDatabase: number
  missingValueRows: number
  importableRows: number
  errors: Array<{ row: number; message: string }>
  preview: Array<Record<string, unknown>>
  sourceTypes: Array<{ value: string; label: string }>
}

export interface HistoryImportBatch {
  id: number
  filename: string
  sourceType: string
  sourceLabel: string
  sourceNote?: string
  platformHint?: string
  totalRows: number
  successCount: number
  duplicateCount: number
  errorCount: number
  missingValueCount: number
  status: string
  remainingRows?: number
  errorsJson?: Array<{ row: number; message: string }>
  createdAt: string
}

export interface RecommendationEffect {
  adoption: {
    totalSchedules: number
    withRecommendation: number
    adoptedCount: number
    adoptionRate: number
    averageDeviationMinutes: number
    deviationBuckets: Array<{ name: string; count: number }>
  }
  comparison: Array<{
    name: string
    scheduleCount: number
    measuredCount: number
    impressions: number
    engagementTotal: number
    engagementRate: number
    avgEngagementRate: number
  }>
  experimentGroups: RecommendationEffect['comparison']
  items: Array<{
    scheduleId: number
    title: string
    platform: Platform
    status: string
    timeSource: TimeSource
    contentType?: string
    scheduledAt: string
    recommendedAt: string | null
    actualPublishAt: string | null
    deviationMinutes: number | null
    executionDelayMinutes: number | null
    sampleCount: number
    impressions: number
    engagementTotal: number
    engagementRate: number
  }>
  notice: string
}

export interface DecisionComponent {
  type: string
  label: string
  weight: number
  rawScore: number
  contribution: number
  description: string
}

/** Full "why this time" chain for one schedule: content → analysis → decision → result. */
export interface PublishDecisionChain {
  scheduleId: number
  title: string
  platform: Platform
  status: string
  explainSource: 'SNAPSHOT' | 'RECOMPUTED'
  explainNotice: string
  content: {
    articleTitle: string | null
    variantTitle: string | null
    contentType: string
    contentTypeName: string
    contentTypeProvider: 'STORED' | 'RULE'
    modelName: string | null
    promptVersion: string | null
    wordCount: number | null
    emojiCount: number | null
    hashtags: string[]
    generationDurationMs: number | null
    tokenUsage: number | null
  }
  analysis: {
    algorithmVersion: string
    weekday: number
    weekdayName: string
    hour: number
    score: number
    confidence: 'HIGH' | 'MEDIUM' | 'LOW'
    weights: { baseline: number; history: number; content: number; timezone: number }
    components: DecisionComponent[]
    hourly: Array<{ hour: number; score: number }>
    bestHour: number
    bestHourScore: number
    accountSampleCount: number
    baselineSampleCount: number
    contentSampleCount: number
    window: { window: string; label: string }
    sourceTypes: Array<{ sourceType: string; label: string; count: number }>
  }
  decision: {
    recommendedAt: string | null
    scheduledAt: string
    actualPublishAt: string | null
    timeSource: TimeSource
    timeSourceLabel: string
    deviationMinutes: number | null
    snapshotScore: number | null
    snapshotConfidence: string | null
    narrative: string | null
    alternatives: RecommendationAlternative[]
  }
  result: {
    sampleCount: number
    impressions: number
    engagementTotal: number
    engagementRate: number | null
    dataSources: string[]
  }
}

export type PublishMode =
  | 'REAL_API'
  | 'DRAFT_ONLY'
  | 'SUBMIT_PUBLISH'
  | 'MANUAL_CONFIRM'
  | 'CDP_PUBLISH'
  | 'MCP_PUBLISH'
  | 'BROWSER_PUBLISH'
  | 'BROWSER_DRAFT'
  | 'WECHATSYNC_CLI'

export type PlatformAccountStatus =
  | 'NOT_CONFIGURED'
  | 'CONNECTING'
  | 'CONNECTED'
  | 'TOKEN_EXPIRED'
  | 'INVALID'
  | 'DISABLED'
  | 'MANUAL_ONLY'
  | 'READY'
  | 'LOGIN_REQUIRED'

export interface PlatformAccount {
  id: number | null
  platform: Platform
  platformName: string
  accountName: string
  authType: 'NONE' | 'OAUTH2' | 'APP_SECRET' | 'QR_LOGIN'
  publishMode: PublishMode
  status: PlatformAccountStatus
  capabilities: string[]
  publishHint: string
  lastTestAt?: string
  lastError?: string
  loginUsername?: string
  lastLoginAt?: string
  sessionDurationSeconds?: number
  appId: string
  clientId: string
  secretConfigured: boolean
  accessTokenConfigured: boolean
  refreshTokenConfigured: boolean
  tokenHint: string
  tokenExpiresAt?: string
  localPublishingEnabled: boolean
  publicPublishEnabled: boolean
  availablePublishModes: PublishMode[]
  config: {
    redirect_uri?: string
    operation_ip?: string
    default_author?: string
    default_cover_media_id?: string
    default_cover_url?: string
    allow_submit_publish?: boolean
    allow_public_publish?: boolean
  }
  connectionGuide: {
    mode:
      | 'OFFICIAL_OAUTH'
      | 'OFFICIAL_OAUTH2_PKCE'
      | 'APP_SECRET'
      | 'MANUAL_ONLY'
      | 'MANUAL_DELIVERY'
      | 'LOCAL_BROWSER_QR'
    consoleUrl: string
    callbackPath?: string
    steps: string[]
  }
  shared?: boolean
}

export interface PublishPackage {
  title: string
  content: string
  hashtags: string[]
  coverImage?: string
  images: string[]
  imageOrder: number[]
  creatorUrl: string
  notice: string
}

export interface LlmConfig {
  provider: string
  baseUrl: string
  apiKey: string
  apiKeyConfigured: boolean
  model: string
  inputPricePerMillion: number
  outputPricePerMillion: number
  currency: 'CNY' | 'USD'
  monthlyBudget: number
  budgetWarningPercent: number
}

export interface LlmConnectionResult {
  connected: boolean
  latencyMs: number
  models: string[]
  message: string
}

export interface LlmUsage {
  days: number
  generations: number
  pricedGenerations: number
  unpricedGenerations: number
  promptTokens: number
  completionTokens: number
  totalTokens: number
  estimatedCost: number
  averageTokens: number
  currency: 'CNY' | 'USD'
  priceConfigured: boolean
  inputPricePerMillion: number
  outputPricePerMillion: number
  monthlyBudget: number
  monthlyCost: number
  budgetWarningPercent: number
  budgetUsedPercent: number
  budgetAlert: boolean
  byModel: Array<{
    model: string
    generations: number
    tokens: number
    cost: number
    unpricedGenerations: number
  }>
  daily: Array<{
    date: string
    tokens: number
    cost: number
    unpricedGenerations: number
  }>
  recent: Array<{
    id: number
    articleTitle: string
    platform: Platform
    model: string
    promptTokens: number
    completionTokens: number
    totalTokens: number
    cost: number | null
    pricingStatus: 'CURRENT_PRICE' | 'SAVED_ESTIMATE' | 'UNPRICED' | 'NO_USAGE'
    createdAt: string
  }>
}

export const platformNames: Record<Platform, string> = {
  WEIBO: '微博',
  XIAOHONGSHU: '小红书',
  WECHAT_OFFICIAL: '微信公众号',
  TOUTIAO: '今日头条',
  X: 'X',
}

/** Extended labels for activity analysis (includes public research datasets). */
export const analysisPlatformNames: Record<string, string> = {
  ...platformNames,
  YOUTUBE: 'YouTube（公开样本）',
}

export const platformColors: Record<Platform, string> = {
  WEIBO: '#f59e0b',
  XIAOHONGSHU: '#ef4444',
  WECHAT_OFFICIAL: '#16a34a',
  TOUTIAO: '#f04438',
  X: '#111827',
}

File: backend/app/schemas/business.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator

Platform = Literal["WEIBO", "XIAOHONGSHU", "WECHAT_OFFICIAL"]


class ArticleCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    source_text: str = Field(min_length=10, max_length=100_000)
    summary: str | None = Field(default=None, max_length=2000)
    topic: str | None = Field(default=None, max_length=100)
    target_audience: str | None = Field(default=None, max_length=150)
    tone: str | None = Field(default="专业自然", max_length=50)
    keywords: list[str] = Field(default_factory=list, max_length=20)


class ArticleUpdate(ArticleCreate):
    status: str | None = None


class GenerateRequest(BaseModel):
    article_id: int
    platforms: list[Platform] = Field(min_length=1)
    style: str = "专业自然"
    length: Literal["SHORT", "MEDIUM", "LONG"] = "MEDIUM"
    target_audience: str | None = Field(default=None, max_length=150)
    include_emoji: bool = True
    include_hashtags: bool = True
    preserve_meaning: int = Field(default=90, ge=50, le=100)
    generation_mode: Literal["QUICK", "DEEP"] = "QUICK"
    creative_goal: str = Field(default="知识分享", min_length=1, max_length=80)

    @field_validator("platforms")
    @classmethod
    def unique_platforms(cls, value: list[Platform]) -> list[Platform]:
        return list(dict.fromkeys(value))


class VariantUpdate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    content_text: str = Field(min_length=1, max_length=100_000)
    hashtags: list[str] = Field(default_factory=list, max_length=20)


class WechatFormatRequest(BaseModel):
    theme: Literal["clean", "brand", "editorial"] = "clean"
    accent_color: str = Field(default="#1677ff", pattern=r"^#[0-9A-Fa-f]{6}$")
    font_size: int = Field(default=16, ge=14, le=20)
    line_height: float = Field(default=1.8, ge=1.4, le=2.2)
    paragraph_spacing: int = Field(default=16, ge=8, le=32)
    first_line_indent: bool = False
    link_footnotes: bool = True


class WechatFormatPreviewRequest(WechatFormatRequest):
    content_text: str = Field(min_length=1, max_length=100_000)


class KeywordRequest(BaseModel):
    article_id: int


class MediaSelectRequest(BaseModel):
    article_id: int
    variant_id: int | None = None
    source: str
    source_id: str | None = None
    image_url: str
    thumbnail_url: str
    photographer_name: str | None = None
    photographer_url: str | None = None
    alt_text: str | None = None
    search_keyword: str | None = None
    usage_type: Literal["COVER", "BODY"] = "BODY"


class RecommendationRequest(BaseModel):
    article_id: int
    variant_id: int | None = None
    platform: Platform
    target_date: date | None = None


class ScheduleCreate(BaseModel):
    article_id: int
    variant_id: int
    account_id: int | None = None
    platform: Platform
    scheduled_at: datetime
    publish_mode: Literal["REAL_API", "DRAFT_ONLY", "MANUAL_CONFIRM"] = "MANUAL_CONFIRM"


class ScheduleUpdate(BaseModel):
    scheduled_at: datetime | None = None
    account_id: int | None = None
    publish_mode: Literal["REAL_API", "DRAFT_ONLY", "MANUAL_CONFIRM"] | None = None


class MetricCreate(BaseModel):
    schedule_id: int
    platform: Platform
    metric_date: date
    impressions: int = Field(default=0, ge=0)
    likes: int = Field(default=0, ge=0)
    comments: int = Field(default=0, ge=0)
    collects: int = Field(default=0, ge=0)
    shares: int = Field(default=0, ge=0)
    followers: int = Field(default=0, ge=0)
    group_type: Literal["RECOMMENDED_TIME", "FIXED_TIME"] = "RECOMMENDED_TIME"
    data_source: Literal["REAL", "MANUAL", "IMPORTED", "SIMULATED"] = "MANUAL"


class ExperimentCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    type: Literal["CONTENT_EFFICIENCY", "PUBLISH_TIME"]
    hypothesis: str = Field(min_length=1)
    start_date: date | None = None
    end_date: date | None = None
    control_description: str | None = None
    treatment_description: str | None = None
    metrics: dict = Field(default_factory=dict)

    @field_validator("end_date")
    @classmethod
    def validate_dates(cls, value: date | None, info):
        start = info.data.get("start_date")
        if value and start and value < start:
            raise ValueError("结束日期不能早于开始日期")
        return value


class ExperimentUpdate(BaseModel):
    name: str | None = None
    hypothesis: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    control_description: str | None = None
    treatment_description: str | None = None
    metrics: dict | None = None
    conclusion: str | None = None


class SettingUpdate(BaseModel):
    value: str = Field(max_length=5000)


class LlmConfigUpdate(BaseModel):
    provider: str = Field(min_length=1, max_length=50)
    base_url: str = Field(default="", max_length=1000)
    api_key: str = Field(default="", max_length=5000)
    model: str = Field(default="", max_length=100)
    input_price_per_million: float = Field(default=0, ge=0, le=1_000_000)
    output_price_per_million: float = Field(default=0, ge=0, le=1_000_000)
    currency: Literal["CNY", "USD"] = "CNY"

    @field_validator("api_key")
    @classmethod
    def normalize_api_key(cls, value: str) -> str:
        return value.strip()

    @field_validator("base_url")
    @classmethod
    def validate_base_url(cls, value: str, info):
        if not value.startswith(("http://", "https://")):
            raise ValueError("接口地址必须以 http:// 或 https:// 开头")
        return value.rstrip("/")

    @field_validator("provider")
    @classmethod
    def reject_local_mock_provider(cls, value: str) -> str:
        normalized = value.strip()
        if normalized.lower() == "mock":
            raise ValueError("不再支持本地模拟模型，请配置真实的大模型服务")
        return normalized


class UserCreate(BaseModel):
    username: str = Field(pattern=r"^[A-Za-z0-9_.-]{3,50}$")
    password: str = Field(min_length=8, max_length=128)
    display_name: str = Field(min_length=1, max_length=100)
    email: str | None = None
    role: Literal["ADMIN", "OPERATOR", "VIEWER"] = "VIEWER"
